### PR TITLE
[SE-0408] Enable nested iteration

### PR DESCRIPTION
**Pull request by simanerush**
# Enables nested iteration for SE-0408

### Changes:
- [x] Cache the generic environment associated with the outer loop in the constraint system and apply it in `getPackElementEnvironment`
- [x] In `mapElementTypeIntoPackContext`, use the interface type to obtain the correct depth.

### CODE REVIEW

**Comment by ahoppen** [comment]
Thanks for also adding `repeat` to the code completion results for loops 🙏🏽

File: include/swift/Sema/ConstraintSystem.h [hunk 1]
```diff
@@ -2301,6 +2301,8 @@ class ConstraintSystem {
   llvm::SmallMapVector<PackElementExpr *, PackExpansionExpr *, 2>
       PackEnvironments;
 
+  llvm::SmallVector<GenericEnvironment * , 4> PackElementGenericEnvironments;
```
Comment (xedin) [inline include/swift/Sema/ConstraintSystem.h:2304]: You'd have to add the same thing to Solution and use SolverScope to control addition, removal of these environments just like we do with i.e. `OpenedExistentialTypes`.

Comment (simanerush) [reply include/swift/Sema/ConstraintSystem.h:2304]: Thank you! I saw that, but I thought that's only needed for maps😅

**Comment by simanerush** [comment]
@swift-ci please smoke test

File: lib/Sema/ConstraintSystem.cpp [hunk 2]
```diff
@@ -789,10 +789,16 @@ ConstraintSystem::getPackElementEnvironment(ConstraintLocator *locator,
...
+  auto genericSignature = outerGenericEnvironment ?
+                  outerGenericEnvironment->getGenericSignature().getCanonicalSignature() :
+                  DC->getGenericSignatureOfContext().getCanonicalSignature();
+
   auto elementSig = ctx.getOpenedElementSignature(
-      DC->getGenericSignatureOfContext().getCanonicalSignature(), shapeParam);
+      genericSignature, shapeParam);
   auto *contextEnv = DC->getGenericEnvironmentOfContext();
-  auto contextSubs = contextEnv->getForwardingSubstitutionMap();
+  auto contextSubs = outerGenericEnvironment ? outerGenericEnvironment->getForwardingSubstitutionMap() : contextEnv->getForwardingSubstitutionMap();
```
Comment (xedin) [inline lib/Sema/ConstraintSystem.cpp:801]: Nit: All the new code here needs to be clang-format'ed

File: lib/Sema/ConstraintSystem.cpp [hunk 3]
```diff
@@ -789,10 +789,16 @@ ConstraintSystem::getPackElementEnvironment(ConstraintLocator *locator,
...
+  GenericEnvironment *outerGenericEnvironment = PackElementGenericEnvironments.empty() ? nullptr : PackElementGenericEnvironments.back();
+
+  auto genericSignature = outerGenericEnvironment ?
+                  outerGenericEnvironment->getGenericSignature().getCanonicalSignature() :
+                  DC->getGenericSignatureOfContext().getCanonicalSignature();
+
   auto elementSig = ctx.getOpenedElementSignature(
-      DC->getGenericSignatureOfContext().getCanonicalSignature(), shapeParam);
+      genericSignature, shapeParam);
   auto *contextEnv = DC->getGenericEnvironmentOfContext();
```
Comment (xedin) [inline lib/Sema/ConstraintSystem.cpp:800]: Instead of dealing with `outerGenericEnvironment` why can't we just do:

```
 auto *contextEnv = PackElementGenericEnvironments.empty() ? PackElementGenericEnvironments.back() : DC->getGenericEnvironmentOfContext();
```

Move `elementSig` down and use `contextEnv.getGenericSignature()...` for it and for `contextSubs` as well?

File: lib/Sema/TypeCheckConstraints.cpp [hunk 4]
```diff
@@ -896,7 +896,7 @@ bool TypeChecker::typeCheckPatternBinding(PatternBindingDecl *PBD,
   return hadError;
 }
 
-bool TypeChecker::typeCheckForEachBinding(DeclContext *dc, ForEachStmt *stmt) {
+bool TypeChecker::typeCheckForEachBinding(DeclContext *dc, ForEachStmt *stmt, GenericEnvironment *packElementEnv) {
```
Comment (xedin) [inline lib/Sema/TypeCheckConstraints.cpp:899]: Nit: This needs formatting as well.

File: lib/Sema/TypeCheckConstraints.cpp [hunk 5]
```diff
@@ -912,7 +912,7 @@ bool TypeChecker::typeCheckForEachBinding(DeclContext *dc, ForEachStmt *stmt) {
     return true;
   };
 
-  auto target = SyntacticElementTarget::forForEachStmt(stmt, dc);
+  auto target = SyntacticElementTarget::forForEachStmt(stmt, dc, false, packElementEnv);
```
Comment (xedin) [inline lib/Sema/TypeCheckConstraints.cpp:915]: Nit: please add `/*ignoreWhereClause=*/` before `false`.

File: lib/Sema/TypeChecker.h [hunk 6]
```diff
@@ -759,7 +759,7 @@ bool typeCheckPatternBinding(PatternBindingDecl *PBD, unsigned patternNumber,
 /// Type-check a for-each loop's pattern binding and sequence together.
 ///
 /// \returns true if a failure occurred.
-bool typeCheckForEachBinding(DeclContext *dc, ForEachStmt *stmt);
+bool typeCheckForEachBinding(DeclContext *dc, ForEachStmt *stmt, GenericEnvironment *packElementEnv);
```
Comment (xedin) [inline lib/Sema/TypeChecker.h:762]: Nit: needs formatting.

**Review by xedin — APPROVED** [review]
Looks good, I left a few minor and refactoring comments inline.

File: lib/Sema/ConstraintSystem.cpp (hunk 3 above)
Comment (simanerush) [reply lib/Sema/ConstraintSystem.cpp:800]: We absolutely can! Thank you for pointing this out🙂

**Comment by simanerush** [comment]
@swift-ci please smoke test

File: lib/Sema/ConstraintSystem.cpp [hunk 7]
```diff
@@ -4403,20 +4405,17 @@ size_t Solution::getTotalMemory() const {
...
+         PackElementGenericEnvironments.size() +
          (DefaultedConstraints.size() * sizeof(void *)) +
          ImplicitCallAsFunctionRoots.getMemorySize() +
-         nodeTypes.getMemorySize() +
-         keyPathComponentTypes.getMemorySize() +
-         size_in_bytes(KeyPaths) +
-         (contextualTypes.size() * sizeof(ASTNode))  +
-         size_in_bytes(targets) +
-         size_in_bytes(caseLabelItems) +
+         nodeTypes.getMemorySize() + keyPathComponentTypes.getMemorySize() +
```
Comment (xedin) [inline lib/Sema/ConstraintSystem.cpp:4411]: I am not sure what happened here but it would be great to remove unrelated formatting changes…

**Comment by xedin** [comment]
Feel free to land this!

File: lib/Sema/ConstraintSystem.cpp (hunk 7 above)
Comment (simanerush) [reply lib/Sema/ConstraintSystem.cpp:4411]: Sorry I missed this😅 I'll remove them

**Comment by simanerush** [comment]
@swift-ci Please smoke test
